osu.Server.Queues.ScoreStatisticsProcessor/Processors/UserTotalPerformanceProcessor.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using Dapper;
+using MySqlConnector;
+using osu.Server.Queues.ScoreStatisticsProcessor.Models;
+
+namespace osu.Server.Queues.ScoreStatisticsProcessor.Processors
+{
+    public class UserTotalPerformanceProcessor : IProcessor
+    {
+        public void RevertFromUserStats(SoloScoreInfo score, UserStats userStats, int previousVersion, MySqlConnection conn, MySqlTransaction transaction)
+        {
+        }
+
+        public void ApplyToUserStats(SoloScoreInfo score, UserStats userStats, MySqlConnection conn, MySqlTransaction transaction)
+        {
+            var items = conn.Query<(int beatmapId, double? pp, double accuracy)>(
+                @"SELECT
+                    s.beatmap_id,
+                    p.pp,
+                    s.data->'$.accuracy'
+                FROM solo_scores s
+                JOIN solo_scores_performance p 
+                ON s.id = p.score_id
+                WHERE s.user_id = @UserId", new
+                {
+                    UserId = userStats.user_id
+                }, transaction);
+
+            // Group items by beatmap_id and extract the item from each group with maximum pp (for both pp and accuracy).
+            // Results are returned in pp-descending order.
+            var filteredItems =
+                items.Where(i => i.pp != null)
+                     .GroupBy(i => i.beatmapId)
+                     .Select(g =>
+                     {
+                         var (_, maxItemPp, maxItemAccuracy) = g.OrderByDescending(i => i.pp).First();
+                         return new
+                         {
+                             BeatmapId = g.Key,
+                             Pp = maxItemPp!.Value,
+                             Accuracy = maxItemAccuracy
+                         };
+                     })
+                     .OrderByDescending(i => i.Pp)
+                     .ToArray();
+
+            // Build the diminishing sum
+            double factor = 1;
+            double totalPp = 0;
+            double totalAccuracy = 0;
+
+            foreach (var item in filteredItems)
+            {
+                totalPp += item.Pp * factor;
+                totalAccuracy += item.Accuracy * factor;
+                factor *= 0.95;
```

Comment:
So this calculates for every score of a user? I wonder if it makes sense to limit to just a user's top 100 or 200, given the diminishing returns. The "weird factor" below would need to then be pulled from a separate count of a user's total unique scores.

Reply:
We're not about to be changing this. This is just following existing implementation which is tried and tested for years (see https://github.com/ppy/osu-performance).